File: 1325. Delete Leaves With a Given Value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def removeLeafNodes(self, root, target, init=True):
        """
        :type root: TreeNode
        :type target: int
        :rtype: TreeNode
        """
        if root == None:
            return
        left = self.removeLeafNodes(root.left, target, False)
        right = self.removeLeafNodes(root.right, target, False)
        if left == 1:
            root.left = None
        if right == 1:
            root.right = None
        if root.left == root.right == None and root.val == target:# leaf
            return None if init else 1# delete the connection between this node and its parent
        return root
            

File: 1325. Delete Leaves With a Given Value.c
/*
Given a binary tree root and an integer target, delete all the leaf nodes with value target.

Note that once you delete a leaf node with value target, 
if its parent node becomes a leaf node and has the value target, 
it should also be deleted (you need to continue doing that until you cannot).
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* _removeLeafNodes(struct TreeNode* root, int target, int depth) {
    if(root == NULL){
        return NULL;
    }
    int l = _removeLeafNodes(root->left, target, depth+1);
    int r = _removeLeafNodes(root->right, target, depth+1);
    if(l==1){
        root->left = NULL;
    }
    if(r==1){
        root->right = NULL;
    }
    // leaf target
    if(root->left == NULL && root->right == NULL && root->val == target){
        return depth == 0 ? NULL : 1;
    }
    return root;
}

struct TreeNode* removeLeafNodes(struct TreeNode* root, int target) {
    return _removeLeafNodes(root, target, 0);
}